### PR TITLE
zapi index create|find commands

### LINKED ISSUE #882: zapi command for archive index search
zapi needs a subcommand that can perform an archive index search, using the api introduced in https://github.com/brimsec/zq/pull/833 .

### PR DESCRIPTION
**Pull request by mattnibs**
Add zar index related commands, allowing users to create and query
indices in archive backed spaces.

Closes #882

### CODE REVIEW

File: cmd/zapi/cmd/index/create.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	Short: "create index on a space",
+	Long: `
+"zapi index create" creates index files in a zar archive using one or more indexing
+rules.
+
+A pattern is either a field name or a ":" followed by a zng type name.
+For example, to index the all fields of type port and the field id.orig_h,
+you would run:
+
+	zapi index create id.orig_h:port
```
Comment (mccanne) [inline cmd/zapi/cmd/index/create.go:27]:
```suggestion
	zapi index create id.orig_h :port
```

File: cmd/zapi/cmd/index/create.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+Each pattern results in a separate microindex file for each log file found.
```
Comment (mccanne) [inline cmd/zapi/cmd/index/create.go:29]:
```suggestion
Each pattern results in a separate, single-key microindex file for each log file found.
```

File: cmd/zapi/cmd/index/create.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+For custom indexes, zql can be used instead of a pattern. This
+requires specifying the key and output file name. For example:
+
+    zapi index create -k id.orig_h -o custom -z "count() by _path, id.orig_h | sort id.orig_h"
+`,
```
Comment (mccanne) [inline cmd/zapi/cmd/index/create.go:35]:
```suggestion
`,
Multiple keys may be specified with multiple -k arguments, in which case the first key is the primary search key, the second key is the secondary search key, and so forth.  For example,

zapi index create -k id.orig_h -k count -o custom -z "count() by _path, id.orig_h | sort id.orig_h,count"

```

File: cmd/zapi/cmd/index/create.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	inputFile  string
+	outputFile string
+	keys       arrayFlag
+	zql        string
+}
+
+func NewCreate(parent charm.Command, f *flag.FlagSet) (charm.Command, error) {
+	c := &CreateCmd{Command: parent.(*IndexCmd).Command}
+	f.Var(&c.keys, "k", "key fields (can be specified multiple times)")
+	f.StringVar(&c.inputFile, "i", "", "input file relative to each zar directory ('' means archive log file in the parent of the zar directory)")
```
Comment (mccanne) [inline cmd/zapi/cmd/index/create.go:55]: I'm not a fan of "" for the parent file.  If we don't like "_", maybe we can come up with something better?  "."?  "THIS"?  Unix find uses "{}" inside of exec.  I'm gonna noodle on this.

File: cmd/zapi/cmd/index/create.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	if err != nil {
+		return err
+	}
+	req := api.IndexPostRequest{
+		Patterns:   args,
+		InputFile:  c.inputFile,
+		OutputFile: c.outputFile,
+	}
+	if c.zql != "" {
+		proc, err := zql.ParseProc(c.zql)
```
Comment (mccanne) [inline cmd/zapi/cmd/index/create.go:75]: cool, so easy!

**Review by mccanne — APPROVED** [review]

File: cmd/zapi/cmd/index/create.go (hunk 4 above)
Comment (mattnibs) [reply cmd/zapi/cmd/index/create.go:59]: I have found it confusing that you need to know about using `_` to do the default behavior you would expect. That said, if we're going to stick with some special character we should probably add a constant somewhere to the code base to make this easier to see/understand.

File: cmd/zapi/cmd/index/create.go (hunk 5 above)
Comment (mattnibs) [reply cmd/zapi/cmd/index/create.go:79]: 💥

File: cmd/zapi/cmd/index/create.go (hunk 4 above)
Comment (mattnibs) [reply cmd/zapi/cmd/index/create.go:58]: Also should we expect the `_` be sent over the api call as well?

**Pull request merged**